### DISCUSSION TITLE
Validation: Use on FrontEnd

### DISCUSSION THREAD

**Discussion by MentalGear** [directus/directus#16561] [Feature Requests]
I wanted to bring this up since your roadmap mentions "Expanding Validation" as a current item.

I'm a huge fan of a unified, universal architecture / design system. Since we are already defining validation rules on the backend, it would be great if we could export those rules (as JS) to make them also available on our frontend.

This should work well for sync rules ( e.g. input is integer ), but async rules probably would need to be adapted to make server calls ( e.g. does username already exist ? ).

Note: [vest](https://github.com/ealush/vest/blob/641324ac0a6b9c8a32a6363a41a9baf8cdf89774/packages/vest/src/core/test/lib/registerTest.ts) does a good job of stashing async validation tests behind sync tests for optimal performance.

**Comment by MentalGear** [comment]
I can't find any details on the Extended Validation item that is on your [current](https://directus.io/road-map/) Roadmap. Could you provide a status on it, and whether the previously mentioned feature to use validation externally (for front-end public apps) would be something that can still be included?

**Comment by rijkvanzanten** [comment]
All the validation rules are stored in either `directus_permissions` or `directus_fields`, which can be read through `/permissions` and `/fields` respectively. The validation rules are in JSON format, so you can reuse those however you want outside of the system 👍🏻

**Reply by MentalGear** [reply]
Thanks for the detailed answer! Do you use a validation library internally that I could also use on the frontend?

**Reply by rijkvanzanten** [reply]
We don't rely on an off-the-shelf validation library. The app/api rely on this util function:

https://github.com/directus/directus/blob/main/packages/shared/src/utils/validate-payload.ts

You can import that from the `@directus/shared` package 🙂

**Reply by MentalGear** [reply]
Thanks @rijkvanzanten for pointing me to the file!

> We don't rely on an off-the-shelf validation library. The app/api rely on this util function:

Well, I don't think there's anything wrong with a good "off the shelf" validation library, mostly because it's often better and more widely tested and feature rich (which might be good if well organized) than custom build ones. Yet, from the code I  noticed that you're using the joi validation library, so I assume I can load the .json validation from /fields into joi? :)

> import Joi from 'joi';

**Reply by rijkvanzanten** [reply]
> Well, I don't think there's anything wrong with a good "off the shelf" validation library, mostly because it's often better and more widely tested and feature rich (which might be good if well organized) than custom build ones

I'm not arguing against that. In our use case, we have some very specific usage requirements that don't exist in the existing popular libraries.

> Yet, from the code I noticed that you're using the joi validation library, so I assume I can load the .json validation from /fields into joi? :)

Yes we use Joi for some aspects in validation, buty no you can't just load in the saved JSON to Joi (for starters because Joi doesn't rely on JSON definitions in the first place). As you can tell from the linked file above, there's a lot more we do than just use Joi directly.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
